File: JavaScript/dfs.js
/**
 * Function that performs depth-first search traversal.
 * @param graph A directed graph represented as an adjency matrix.
 * @param root The initial node to start the traversal
 */
function dfs(graph, root){
    // initializes relevant data structures
    toVisit = [root];        // stack initialized with root node
    output = [];             // output to be returned
    visited = new Set();     // set that tracks already visited node
    
    // iterate over each node to be visited
    while(toVisit.length > 0){
        // pop the node from the toVisit stack
        node = toVisit.pop();
        // check if node is visited
        if (!visited.has(node)){
            output.push(node); // add node to the output
            visited.add(node); // mark node as visited       
            // iterate over the children (in reverse order)
            let children = graph[node];
            for(let i = children.length - 1 ; i >= 0 ; i--){
                toVisit.push(children[i]);
            }
        }
    }

    return output;
}



assert = require('assert');




// assert.deepStrictEqual: this method from the assert module verifies whether the two arrays have the same element values.
// That is, it does a 'deep' equality check (ie, values within the array)
// if the test fails, it throws an error


// test case 1
graph = {"+": ["*",3], "*":[2,7], 2:[],7:[],3:[]};
root = "+" ;
assert.deepStrictEqual(dfs(graph, root), ['+', '*', 2, 7, 3])

// test case 2
graph = {0: [1,3], 1:[2,3], 2:[3,1], 3:[0,1]};
root = 0 ;
assert.deepStrictEqual(dfs(graph, root), [0, 1, 2, 3])

File: Java/datastructures/DFS.java
package datastructures;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Arrays;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.TreeSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.Stack;

public class DFS {
    public List<String> traverse(String root, Map<String,List<String>> graph){
        // queue with nodes to be visited (initialized with root node)
        Stack<String> toVisit = new Stack() ;
        toVisit.add(root);
        // list of visited nodes
        List<String> output = new ArrayList();
        // set of visited nodes, to avoid getting stuck
        Set<String> visited = new HashSet();

        // keep visiting while there are nodes in the queue
        while(!toVisit.isEmpty()){
            String node = toVisit.pop();
            // check node is not visited
            if(!visited.contains(node)){
                output.add(node);
                // mark node as visited
                visited.add(node);

                // iterate over node's children and add back to queue
                List<String> children = graph.get(node);
                for(int i = children.size() - 1; i >= 0 ; i--){
                    toVisit.add(children.get(i));
                }
            }
        }


        // returns the visited nodes, in order
        return output;
    }

    
}